### PR TITLE
[Bugfix] Use realistic top_k in _dummy_sampler_run to avoid SM120 startup hang

### LINKED ISSUE #42987: [SM120] _dummy_sampler_run hangs indefinitely on RTX 5090 due to top_k=vocab_size-1 triggering an SM120-broken top-k masking kernel (one-line fix)
### Summary

On physical RTX 5090 hardware (SM120 / compute_120f), `vllm serve` hangs indefinitely during engine startup. The hang is in `_dummy_sampler_run` (called from `profile_run` → `_initialize_kv_caches`) when it invokes the sampler with `top_k = logits.size(1) - 1` (i.e. `vocab_size - 1`, which is **151935** for `Qwen/Qwen2.5-3B-Instruct`).

Real production sampling uses small `top_k` values (e.g. 40, 50) and does not hang. Substituting a small `top_k` in the dummy run resolves the startup hang with no functional regression — the dummy sampler exists only for KV-cache memory profiling, where realistic top_k values give an identical memory footprint.

### One-line proposed fix

`vllm/v1/worker/gpu_model_runner.py`, in `_dummy_sampler_run`:

```diff
- top_k=dummy_tensors(logits.size(1) - 1),
+ top_k=dummy_tensors(50),
```

(or any small constant in the typical-production range, e.g. 20–100.)

I'd be glad to send a PR. Wanted to surface the diagnosis first in case there's a different preferred remediation (e.g. detect SM120 and short-circuit, or fix the underlying kernel).

### Diagnostic evidence (py-spy, captured 2026-05-17 on a real RTX 5090)

```
Thread MainThread (active): "VLLM::EngineCore"
    apply_top_k_top_p_triton (vllm/v1/sample/ops/topk_topp_triton.py:1000)
    apply_top_k_top_p          (vllm/v1/sample/ops/topk_topp_sampler.py:252)
    forward_native             (vllm/v1/sample/ops/topk_topp_sampler.py:106)
    sample                     (vllm/v1/sample/sampler.py:276)
    forward                    (vllm/v1/sample/sampler.py:97)
    _dummy_sampler_run         (vllm/v1/worker/gpu_model_runner.py:5663)
    profile_run                (vllm/v1/worker/gpu_model_runner.py:5855)
    determine_available_memory (vllm/v1/worker/gpu_worker.py:370)
    _initialize_kv_caches      (vllm/v1/engine/core.py:250)
```

`py-spy dump --native` shows the host thread is stuck inside `cuLaunchKernel → clock_gettime` — a `CUDA_LAUNCH_BLOCKING=1` sync poll waiting for a GPU kernel that **never returns**. The process sits at 100% CPU, GPU at 0% util, allocated memory ~7128 MiB, observed for >9 hours before manual `SIGTERM` cleanup.

**The same hang reproduces with `VLLM_USE_FLASHINFER_SAMPLER=1`**:

```
top_k_mask_logits          (flashinfer/sampling.py:564)
top_k_mask_logits          (flashinfer/sampling.py:1820)
top_k_top_p_sampling_from_logits (flashinfer/sampling.py:1414)
flashinfer_sample          (vllm/v1/sample/ops/topk_topp_sampler.py:391)
forward_cuda               (vllm/v1/sample/ops/topk_topp_sampler.py:140)
sample                     (vllm/v1/sample/sampler.py:276)
_dummy_sampler_run         (vllm/v1/worker/gpu_model_runner.py:5663)
...
```

So **both the Triton-based and the FlashInfer-based top-k masking kernels hang for this `top_k` value on SM120.** Switching between the two doesn't help; the bug appears to be specific to the codegen/scheduling these kernels emit when `top_k` approaches `vocab_size`.

### Verified workaround

I monkey-patched `_dummy_sampler_run` to substitute `top_k=50` into the dummy `SamplingMetadata` (via a `.pth` injection in the venv's site-packages, gated on an env var so it's opt-in). With the patch active, `vllm serve` reaches `Application startup complete` in ~5 seconds. No regressions observed in subsequent inference behavior (other independent SM120 issues exist; see "Other notes" below).

### Why the dummy-run argument is the worst case for these kernels

`_dummy_sampler_run` at `gpu_model_runner.py` line 5648 (vLLM 0.20.1) creates:

```python
dummy_metadata = SamplingMetadata(
    temperature=dummy_tensors(0.5),
    all_greedy=False,
    all_random=False,
    top_p=dummy_tensors(0.9),
    top_k=dummy_tensors(logits.size(1) - 1),  # <-- vocab_size - 1
    ...
)
```

`logits.size(1)` is the model's vocab size. For Qwen2.5-3B this is 151,936, so `top_k` becomes 151,935 — essentially "keep top-151,935 of 151,936". This is presumably intended as a worst-case argument for memory profiling, but it exercises a kernel codegen path that doesn't terminate on SM120 in either implementation.

Real production requests pass per-call top_k from the SamplingParams (typically ≤ 100). Substituting a realistic value in the dummy run keeps the memory-profiling intent intact while avoiding the buggy code path.

### Hardware and software

| Item | Value |
|---|---|
| GPU | NVIDIA GeForce RTX 5090 (SM120 / compute_120f, 32 GB VRAM) |
| OS | Ubuntu 24.04.4 LTS, kernel 6.17.0-22-generic |
| NVIDIA driver | 595.58.03 (open kernel modules) |
| CUDA toolkit | 13.2 |
| torch | 2.11.0+cu130 |
| vllm | 0.20.1 (PyPI wheel) |
| flashinfer-python | 0.6.11 (PyPI wheel, JIT-only) |
| Env | `CUDA_LAUNCH_BLOCKING=1`, `FLASHINFER_DISABLE_VERSION_CHECK=1`, `--attention-backend flashinfer` |

Not WSL2, not DGX Spark / SM121 — physical bare-metal RTX 5090.

### Other notes (not blocking this fix)

After applying the workaround, two additional SM120 hangs become visible later in the lifecycle:

1. With `--enforce-eager`: first inference request hangs in `default_unquantized_gemm` at `vllm/model_executor/layers/utils.py:98` (the lm-head matmul). Same shape: kernel launched, host blocks on CLB=1 sync, GPU never returns.
2. With `cudagraph_mode=FULL_AND_PIECEWISE`: weight loading stalls and the GPU enters `[requires reset]` at 6752 MiB allocated.

These appear to be independent kernel-scheduling issues on SM120 affecting different kernel families. The dummy_sampler fix is independent and worth landing on its own; the others I plan to file separately (or comment on existing trackers) once I have similar diagnostic evidence.

### Reproduction

Any vLLM 0.20.1 (likely 0.21.x too — not retested with the fix) startup on an RTX 5090 with default sampling settings:

```
CUDA_LAUNCH_BLOCKING=1 FLASHINFER_DISABLE_VERSION_CHECK=1 \
vllm serve Qwen/Qwen2.5-3B-Instruct \
  --enforce-eager \
  --attention-backend flashinfer \
  --gpu-memory-utilization 0.85 \
  --max-model-len 2048 \
  --port 8000
```

Hangs indefinitely at `profile_run` step. With `top_k=50` substituted in `_dummy_sampler_run`, reaches `Application startup complete` in ~5 seconds.

### Offer

Reliable RTX 5090 reproducer available. Happy to test the official patch (or any alternative) on this hardware with overnight turnaround.

### PR DESCRIPTION
**Pull request by daniel-devlab**
Purpose
Fixes #42987.
On RTX 5090 (SM120), vllm serve hangs indefinitely during engine startup. As diagnosed in the issue with py-spy, the hang is in _dummy_sampler_run (profile_run → _initialize_kv_caches): it builds dummy SamplingMetadata with top_k = logits.size(1) - 1 (i.e. vocab_size - 1), and the top-k masking kernel never returns for that value on SM120. The issue reports the same hang in both the Triton path (apply_top_k_top_p_triton) and the FlashInfer path (top_k_mask_logits).

_dummy_sampler_run exists only to warm up / memory-profile the sampler for KV-cache sizing. Its inputs are already entirely synthetic (torch.rand_like hidden states, dummy temperature/top_p/penalties), and its output is discarded. The vocab_size - 1 value is an arbitrary "max-ish" placeholder — nothing depends on it, and the top-k masking operates on the full [num_reqs, vocab_size] logits tensor regardless, so the memory footprint is independent of top_k.

This PR substitutes a realistic top_k (50, within the typical production range of ≤100) in the dummy run. This avoids the buggy SM120 kernel codegen path with no functional change and an identical memory-profiling footprint. With this change, the issue's reproducer reaches Application startup complete in ~5s instead of hanging.

This is a targeted fix for the startup hang. The underlying SM120 top-k kernel issue, which reproduces in two independent implementations and likely needs an upstream fix in NVIDIA's / Triton's SM120 codegen  is out of scope here and is best tracked separately, as discussed in the issue.

Test Plan

pre-commit run --files vllm/v1/worker/gpu_model_runner.py ==>  all hooks pass.
The reporter (@dclarktandem) has offered to verify on physical RTX 5090 hardware: a clean-GPU run of the issue's reproducer command against the patched build, confirming both Application startup complete and a coherent inference response.

Test Result
pre-commit passes locally. Hardware verification on RTX 5090 to follow on this PR.

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -5892,7 +5892,12 @@ def _dummy_sampler_run(
             all_greedy=False,
             all_random=False,
             top_p=dummy_tensors(0.9),
-            top_k=dummy_tensors(logits.size(1) - 1),
+            # Use a realistic top_k value. The dummy run only exists for
+            # KV-cache memory profiling, and the memory footprint is
+            # independent of top_k. A near-vocab_size value (vocab_size - 1)
+            # triggers an SM120 top-k masking kernel hang on RTX 5090 in both
+            # the Triton and FlashInfer paths; see issue #42987.
+            top_k=dummy_tensors(50),
```
[the remark this answers is not in the document]
Comment (daniel-devlab) [reply vllm/v1/worker/gpu_model_runner.py]: Good catch on the small-vocab edge case. I've capped at vocab_size - 1 (min(50, logits.size(1) - 1)) rather than min(50, logits.size(1) // 2) — the // 2 isn't needed for any real vocab size, and logits.size(1) is always ≥ 1 here so max(1, ...) is redundant. This keeps it minimal while preventing top_k from exceeding vocab size.

**Comment by dclarktandem** [comment]
Verified on RTX 5090 (SM120, driver 595.58.03, CUDA 13.2). One caveat upfront: the change was cherry-picked onto vLLM 0.20.1 rather than tested from main. Our dev venv is pinned to 0.20.1 because vLLM main has a separate SM120 regression (hang in `vocab_parallel_embedding` during `profile_run`, even with `--enforce-eager`) that prevents a clean direct-on-main verification. The PR's diff applies to 0.20.1 cleanly — the surrounding 5 lines of `_dummy_sampler_run` are identical between the two versions; only the line number differs (5647 in 0.20.1 vs 5895 in main). Diff against the backup matches the PR byte-for-byte.

**Result for this PR's scope:** ✓ Fix confirmed.

- Without the patch: `_dummy_sampler_run` hangs indefinitely during `profile_run`, as documented in #42987.
- With the patch (`top_k=dummy_tensors(min(50, logits.size(1) - 1))`): on a clean GPU after a cold boot, the engine reaches `INFO: Application startup complete` in ~30 seconds.

Issue reproducer command was used unmodified other than the locally-required `CUDA_LAUNCH_BLOCKING=1` and `FLASHINFER_DISABLE_VERSION_CHECK=1` that were already in the original report.

**Out of scope but flagged for transparency:** Inference past startup is currently blocked in our environment by a separate SM120 issue in the lm-head matmul (the underlying cuBLAS GEMM also hangs on SM120 here, and our local Triton-based workaround is producing corrupted logits). This is unrelated to `_dummy_sampler_run` and unaffected by this PR — we tracked it independently before this issue was filed. Mentioning it only so the verification picture is complete; it does not change the assessment of this PR.

Thanks for the quick turnaround on this.

**Comment by daniel-devlab** [comment]
Quick status summary for any reviewer:

Targeted one-line fix for #42987 (SM120 startup hang in _dummy_sampler_run).
Diff is +7/-1, contained to gpu_model_runner.py, with comment explaining why the change is footprint-neutral for KV-cache memory profiling.
Review feedback from @gemini-code-assist on tiny-vocab safety addressed: top_k=dummy_tensors(min(50, logits.size(1) - 1)).
RTX 5090 hardware verification by @dclarktandem confirms the fix — patched build reaches Application startup complete; unpatched hangs (see verification comment above).
pre-commit passes; DCO signed.

Would appreciate a review when someone has a moment.
